src/framegrab/grabber.py

```diff
@@ -703,44 +684,57 @@ def _substitute_rtsp_password(self, config: dict) -> dict:
 
         return config
 
+    def _apply_camera_specific_options(self, options: dict) -> None:
+        if options.get("resolution"):
+            camera_name = self.config.get("name", "Unnamed RTSP Stream")
+            raise ValueError(f"Resolution was set for {camera_name}, but resolution cannot be set for RTSP streams.")
+
+    def _open_connection(self):
+        self.capture = cv2.VideoCapture(self.stream)
+        if not self.capture.isOpened():
+            raise ValueError(
+                f"Could not open RTSP stream: {self.stream}. Is the RTSP URL correct? Is the camera connected to the network?"
+            )
+        logger.debug(f"Initialized video capture with backend={self.capture.getBackendName()}")
+
+    def _close_connection(self):
+        with self.lock:
+            self.capture.release()
+
     def grab(self) -> np.ndarray:
+        if not self.keep_connection_open:
+            self._open_connection()
+
         with self.lock:
-            (
-                ret,
-                frame,
-            ) = self.capture.retrieve()  # grab and decode since we want this frame
+            ret, frame = self.capture.retrieve() if self.keep_connection_open else self.capture.read()
             if not ret:
                 logger.error(f"Could not read frame from {self.capture}")
 
-        if frame is None:
-            camera_name = self.config.get("name", "Unnamed RTSP Stream")
-            raise ValueError(f"Could not grab a frame from {camera_name}. Is the RTSP stream connected?")
-        frame = self._process_frame(frame)
+        if not self.keep_connection_open:
+            self._close_connection()
```

Comment:
Closing the connection should be inside a `finally:` block, so that we close the connection if any exception gets raised.  So I'd refactor it like this:

```
def grab()
    if self.keep_connection_open:
        self._open_connection()
        try:
            return self._grab_open()
        finally:
            self.close_connection()
    else:
        return self._grab_open()
````

and move the guts into `_grab_open`

Reply:
Done.